## SLE712 Assignment 3 (Trimester 1, 2020)
This project is an assessment for Practical 3 of SLE712 (Bioinformatics and Molecular Biology Techniques) taught by Dr. Mark Ziemann of the School of Life and Environmental Sciences of Deakin University, Waurn Ponds Campus on the first trimester of 2020. This project uses the R language in manipulating genetic data to gain insights and develop data analysis skills necessary for Bioinformatics. This project uses Github as versioning control.

## Contents
* Folder Structure
* Getting Started
* Authors
* License
* Acknowledgments

## Folder Structure

The project comprises of three main folders in which you would store certain files.

### Main Folder

The main folder contains the following subfolders:
- Data
- ProjectDetails
- Scripts
which will be discussed in detail in the following sections.

#### README files
This project contains two types of README files which include:
- **The overall project README file** which the reader is currently reading.
- **Part 1 and 2 READMEs** which should contain the descriptions and the purpose of each script used in the project together with their inputs and outputs.

The README files have their corresponding HTML files for ease of reading.

### Data/

This folder should contain data files like .csv, .tsv, .fa, etc.

In this folder:

- A gene expression data and a growth data for part 1 of the project taken from [Dr. Mark Ziemann's Github repository for the unit](https://github.com/markziemann/SLE712_files/tree/master/bioinfo_asst3_part1_files).
- A sample gene sequence file for part 2 of the project taken from [Dr. Mark Ziemann's Github repository for the unit](https://github.com/markziemann/SLE712_files/tree/master/bioinfo_asst3_part2_files).
- Files generated by the RScripts like downloads, images, and files pertaining to the creation of BLAST databases. 

### Scripts/

This folder should contain R script files.

In this folder:

- Script for part 1 of the project containing basic data manipulation on a gene expression data and a tree growth data 
- Script for part 2 of the project containing analysis on a sample gene sequence against a library of *Escherichia coli* genes

### ProjectDetails/

This folder should contain all files that describes the project, the project report, and the files rendered by the scripts.

In this folder:

- **Assignment Details** which should contain the description of the assignment including the questions, the list of outputs required, and the rubric authored by the unit chair, Dr. Mark Ziemann.

- **Project Report** which should contain the answers to the questions from the assignment details. This currently includes:
  - Word doc, "SLE712_A3_Report.docx"

## Getting Started

These instructions will get you a copy of the project up and running on your local machine.  

### Prerequisites

Things you need to install

- R version 3.6.3 (2020-02-29)
- RStudio Version 1.2.1335
- Packages and Libraries 
  *For part 2, make sure to install these packages before you run the code*
  + seqinr
  + R.utils
  + rBLAST
- A function (made by Dr. Ziemann) referenced from this [link](https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R)

### Installing

#### R, RStudio, and Packages Installation Tutorial
To install R, RStudio, and Packages, go to [this tutorial from datacamp](https://www.datacamp.com/community/tutorials/installing-R-windows-mac-ubuntu)

### Cloning the repository
To clone the repository, follow this [tutorial from RStudio](https://support.rstudio.com/hc/en-us/articles/200532077-Version-Control-with-Git-and-SVN) and use this link as the **repository URL:**

  *<https://github.com/megan0012/SLE712-Assignment-3.git>*

## Authors

* **Megan Soria** - [LinkedIn Profile](https://www.linkedin.com/in/megan-soria-b8a97857/)
* **Ankush Dehlia** - [LinkedIn Profile](https://www.linkedin.com/in/ankush-dehlia/)


## License

This project is licensed under the GNU GENERAL PUBLIC LICENSE - see the LICENSE.md file for details.





